src/server.rs

```diff
@@ -0,0 +1,346 @@
+        self.channel.send(ChannelMessage::Gossip(addr.into()))?;
+        Ok(())
+    }
+}
+
+/// UDP server for ScuttleButt communication.
+struct UdpServer {
+    channel: UnboundedReceiver<ChannelMessage>,
+    scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    socket: Arc<UdpSocket>,
+}
+
+impl UdpServer {
+    async fn new(
+        channel: UnboundedReceiver<ChannelMessage>,
+        scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    ) -> anyhow::Result<Self> {
+        let socket = {
+            let address = &scuttlebutt.lock().await.self_node_id;
+            Arc::new(UdpSocket::bind(address).await?)
+        };
+
+        Ok(Self {
+            scuttlebutt,
+            channel,
+            socket,
+        })
+    }
+
+    /// Listen for new ScuttleButt messages.
+    async fn run(&mut self) -> anyhow::Result<()> {
+        let mut heartbeat_interval = time::interval(MAX_HEARTBEAT_DELTA);
+        let mut gossip_interval = time::interval(GOSSIP_INTERVAL);
+        let mut rng = SmallRng::from_entropy();
+
+        let mut buf = [0; BUF_SIZE];
+        loop {
+            tokio::select! {
+                result = self.socket.recv_from(&mut buf) => match result {
+                    Ok((len, addr)) => {
+                        let _ = self.process_package(addr, &buf[..len]).await;
+                    }
+                    Err(err) => return Err(err.into()),
+                },
+                _ = heartbeat_interval.tick() => self.reap_zombies().await,
+                _ = gossip_interval.tick() => self.gossip_multiple(&mut rng).await,
+                message = self.channel.recv() => match message {
+                    Some(ChannelMessage::Gossip(addr)) => {
+                        let _ = self.gossip(addr).await;
+                    },
+                    Some(ChannelMessage::Shutdown) | None => break,
+                },
+            }
+        }
+        Ok(())
+    }
+
+    /// Process a single UDP packet.
+    async fn process_package(&self, addr: SocketAddr, data: &[u8]) -> anyhow::Result<()> {
+        // Handle gossip from other servers.
+        let message = bincode::deserialize(data)?;
+        let response = self.scuttlebutt.lock().await.process_message(message);
+
+        // Send reply if necessary.
+        if let Some(message) = response {
+            let message = bincode::serialize(&message)?;
+            self.socket.send_to(&message, addr).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Remove all nodes without an up-to-date heartbeat.
+    async fn reap_zombies(&self) {
+        let mut scuttlebutt = self.scuttlebutt.lock().await;
+        let nodes = scuttlebutt.cluster_state_map.node_states();
+        let now = Self::secs_since_epoch();
+
+        nodes.retain(|_node_id, node_state| {
+            let heartbeat = node_state
+                .get(HEARTBEAT_KEY)
+                .and_then(|value| value.parse().ok());
+            now - heartbeat.unwrap_or(0) <= MAX_HEARTBEAT_DELTA.as_secs()
+        });
```

Comment:
This is what I'd consider a "naive" approach to heartbeats: Remove all nodes that exceed a specific duration.

However, looking at the `compute_delta` method it seems like this could behave rather strangely with gossiping, since nodes aren't removed on gossip. So every node would have to remove it themselves and it would be possible that one node removes the node, another gossips about it because its timer runs slightly differently and then the original node suddenly has the dead node again.

Sooner or later all nodes will remove this themselves, but I'm not sure if a "proper" solution wouldn't filter dead nodes *before* gossiping in addition to periodically? Or just before gossiping in general since that happens periodically? I haven't read through the Cassandra source code, but I'd be surprised if they just roll with this issue.

Reply:
We never remove nodes.

Each just separately assesses whether they think a node is dead or not. 
This information is not shared with other nodes, but an internal client in the node can get the list of live nodes.

Reply:
Ah okay, this probably answers a question I forgot to pose: It's not necessary to have any push-based API informing consumers about node deaths, right? So the consumers of the API just query for which nodes are alive whenever they need that information.

Reply:
That's a good question.
We can do a pull API only for the moment I think.